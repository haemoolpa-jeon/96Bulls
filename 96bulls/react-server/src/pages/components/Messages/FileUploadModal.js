import React, { Component } from 'react';
import Modal from 'react-modal';
import Input from '../common/Input';
import Button from '../common/Button';
import mime from 'mime-types';
import PropTypes from 'prop-types';
import styles from './FileUplodeModal.module.scss';

class FileUploadModal extends Component {
  state = {
    file: null,
    authorized: ['image/jpeg', 'image/png'],
  };

  addFile = (event) => {
    const file = event.target.files[0];
    if (file) {
      this.setState({ file });
    }
  };

  sendFile = () => {
    const { file } = this.state;
    const { uploadFile, closeModal } = this.props;
    if (file !== null) {
      if (this.isAuthorized(file.name)) {
        const metadata = { contentType: mime.lookup(file.name) };
        uploadFile(file, metadata);
        closeModal();
        this.clearFile();
      }
    }
  };

  /**
   * @param filename
   * @return boolean
   */
  isAuthorized = filename => this.state.authorized.includes(mime.lookup(filename));

  clearFile = () => this.setState({ file: null });

  render() {
    const customStyles = {
      content: {
        top: '50%',
        left: '50%',
        right: 'auto',
        bottom: 'auto',
        width: '60rem',
        marginRight: '-50%',
        transform: 'translate(-50%, -50%)',
        border: 'none',
        background: '#eee',
        padding: '3rem',
        boxShadow: '0px 1px 5px 0px rgba(0, 0, 0, 0.2), 0px 2px 2px 0px rgba(0, 0, 0, 0.14), 0px 3px 1px -2px rgba(0, 0, 0, 0.12)',
      },
    };

    const { isOpen, closeModal } = this.props;
    return (
      <Modal
        shouldCloseOnOverlayClick
        isOpen={isOpen}
        onRequestClose={closeModal}
        style={customStyles}
        ariaHideApp={false}
      >
        <h1>Select Image File</h1>
        <Input
          id="file"
          name="file"
          type="file"
          placeholder="File types: jpg, png"
          onChange={this.addFile}
        />
        <div className={styles['button-wrapper']}>
          <Button onClick={this.sendFile}>Send</Button>
        </div>
      </Modal>
    );
  }
}

FileUploadModal.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  closeModal: PropTypes.func.isRequired,
  uploadFile: PropTypes.func.isRequired,
};

export default FileUploadModal;
