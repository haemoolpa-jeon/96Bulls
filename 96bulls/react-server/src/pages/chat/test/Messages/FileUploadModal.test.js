import React from 'react';
import FileUploadModal from '../../components/Messages/FileUploadModal';
import {shallow} from 'enzyme/build';

describe('FileUploadModal', () => {
  let component = null;

  const isOpen = false;
  const mockCloseModal = jest.fn(e => e);
  const mockUploadFile = jest.fn(e => e);
  it('초기 렌더링이 문제없이 되야함', () => {
    component = shallow(<FileUploadModal
      uploadFile={mockUploadFile}
      closeModal={mockCloseModal}
      isOpen={isOpen}
    />);
  });

  it('.jpg 파일은 true를 리턴해야 함', () => {
    const instance = component.instance()
    expect(instance.isAuthorized('6ff826f6ed144c29ab2ed42f310f4612_20181016181705.jpg')).toBe(true)
  });

  it('.svg 파일은 false 리턴해야 함', () => {
    const instance = component.instance()
    expect(instance.isAuthorized('img.svg')).toBe(false)
  });

  it('초기 렌더링 스냅샷 일치함', () => {
    expect(component).toMatchSnapshot();
  });
});
