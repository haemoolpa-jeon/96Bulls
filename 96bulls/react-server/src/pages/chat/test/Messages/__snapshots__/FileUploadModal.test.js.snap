// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileUploadModal 초기 렌더링 스냅샷 일치함 1`] = `
<Modal
  ariaHideApp={false}
  bodyOpenClassName="ReactModal__Body--open"
  closeTimeoutMS={0}
  isOpen={false}
  onRequestClose={[MockFunction]}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  role="dialog"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
  style={
    Object {
      "content": Object {
        "background": "#eee",
        "border": "none",
        "bottom": "auto",
        "boxShadow": "0px 1px 5px 0px rgba(0, 0, 0, 0.2), 0px 2px 2px 0px rgba(0, 0, 0, 0.14), 0px 3px 1px -2px rgba(0, 0, 0, 0.12)",
        "left": "50%",
        "marginRight": "-50%",
        "padding": "3rem",
        "right": "auto",
        "top": "50%",
        "transform": "translate(-50%, -50%)",
        "width": "60rem",
      },
    }
  }
>
  <h1>
    이미지 파일을 선택하세요
  </h1>
  <Input
    id="file"
    name="file"
    onChange={[Function]}
    placeholder="File types: jpg, png"
    type="file"
  />
  <div
    className="button-wrapper"
  >
    <Button
      onClick={[Function]}
    >
      전송
    </Button>
  </div>
</Modal>
`;
